decomp/semantics/uds.py

```diff
@@ -960,37 +1188,199 @@ def _add_edge_annotation(self, edge, attrs):
+                       sentence_ids: Dict[str, str], name: str = 'UDS') -> 'UDSDocument':
+        """Construct a UDSDocument from a dictionary
 
-        Find and return the unique root (source) node of a digraph.  If
-        the digraph has no source nodes, or more than one source node,
-        raise an Exception.
+        Since only the document graphs are serialized, the sentence
+        graphs must also be provided to this method call in order
+        to properly associate them with their documents.
 
         Parameters
         ----------
-        digraph
-            digraph to search
+        document
+            a dictionary constructed by networkx.adjacency_data,
+            containing the graph for the document
+        sentence_graphs
+            a dictionary containing (possibly a superset of) the
+            sentence-level graphs for the sentences in the document
+        sentence_ids
+            a dictionary containing (possibly a superset of) the
+            UD sentence IDs for each graph
+        name
+            identifier to append to the beginning of node ids
         """
+        document_graph = UDSDocumentGraph.from_dict(document, name)
+        sent_graph_names = set(map(lambda node: node['semantics']['graph'], document['nodes']))
+        sent_graphs = {gname: sentence_graphs[gname] for gname in sent_graph_names}
+        sent_ids = {gname: sentence_ids[gname] for gname in sent_graph_names}
+        genre = name.split('-')[0]
+        timestamp = cls._get_timestamp_from_document_name(name)
+        return cls(sent_graphs, sent_ids, name, genre, timestamp, document_graph)
+
+    @staticmethod
+    def _get_timestamp_from_document_name(document_name):
+        timestamp = re.search('\d{8}_?\d{6}', document_name)
+        return timestamp[0] if timestamp else None
+
+    def add_sentence_graphs(self, sentence_graphs: Dict[str, UDSSentenceGraph], 
+                                  sentence_ids: Dict[str, str]) -> None:
+        """Add additional sentences to a document
 
-        root_candidates = [node for (node, degree) in digraph.in_degree() if degree == 0]
-        if len(root_candidates) != 1:
-            raise Exception('no unique root of digraph {}'.format(digraph.name))
-        return root_candidates[0]
+        Parameters
+        ----------
+        sentence_graphs
+            a dictionary containing the sentence-level graphs
+            for the sentences in the document
+        sentence_ids
+            a dictionary containing the UD sentence IDs for each graph
+        name
+            identifier to append to the beginning of node ids
+        """
+        for gname, graph in sentence_graphs.items():
+            sentence_graphs[gname].sentence_id = sentence_ids[gname]
+            sentence_graphs[gname].document_id = self.name
+            self.sentence_graphs[gname] = graph
+            self.sentence_ids[gname] = sentence_ids[gname]
+            for node in graph.semantics_nodes:
```

Comment:
To be clear, this implies that we're creating a document node for every semantics node, correct? I believe this is right; I just hadn't thought about it before, since I'd mainly been thinking about just the nodes that have annotations. But this is more principled. It just means we'll have a lot of nodes floating out on their own---not hooked into the graph. Then, we just add edges where we have annotations?

Reply:
I deliberated about this for a little while and in the end just decided to create a document node for every semantics node—and yes, largely out of some vague sense that it was more "principled," even though I can't immediately think of a case where we would be _harmed_ by just creating them dynamically upon annotation.

Reply:
The sense in which I think it's more principled is that I think we want to avoid adding graph structure while adding annotations where possible. And if we do, we want to allow the user not to choose whether to add that structure---as in [`UDSGraph._add_node_annotation`](https://github.com/decompositional-semantics-initiative/decomp/blob/master/decomp/semantics/uds.py#L828). The reason that we sometimes add structure has to do with UDS annotations collected before predpatt, where we used predicate- and argument-finding rules that genrally matched with predpatt predicates and arguments, but not always. So the question is whether to take the union of existing nodes and nodes implied by those annotations or the intersection.

Here, you're right that it probably won't matter for nodes, since now all the annotations are collected on structure that is systematically related to predpatt structure, though for edges, we still are going to need to add structure as part of adding annotations, since I don't think there's a principled way of determining which edges to include in the document graph. Those don't come from any existing extraction system.